atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/TransactionManagerTest.java

```diff
@@ -168,6 +170,74 @@ public void shouldNotConflictIfImmutableTimestampLockExpiresIfNoReadsOrWrites()
         txnManagerWithMocks.runTaskThrowOnConflict(txn -> null);
     }
 
+    @Test
+    public void writeDoesNotPopulateTimestampCache() {
+        Transaction transaction = txMgr.runTaskWithRetry(txn -> {
+            put(txn, TEST_TABLE, "row", "column", "test");
+            return txn;
+        });
+
+        assertThat(timestampCache.getCommitTimestampIfPresent(transaction.getTimestamp())).isNull();
+    }
+
+    @Test
+    public void readCachesWritersStartToCommitTimestamp() {
+        Transaction writer = txMgr.runTaskWithRetry(txn -> {
+            put(txn, TEST_TABLE, "row", "column", "test");
+            return txn;
+        });
+
+        assertThat(timestampCache.getCommitTimestampIfPresent(writer.getTimestamp())).isNull();
+
+        txMgr.runTaskWithRetry(txn -> {
+            get(txn, TEST_TABLE, "row", "column");
+            return txn;
+        });
+
+        assertThat(timestampCache.getCommitTimestampIfPresent(writer.getTimestamp()))
+                .isNotNull()
+                .isEqualTo(transactionService.get(writer.getTimestamp()));
+    }
+
+    @Test
+    public void overwriteCachesPreviousWritersStartToCommitTimestamp() {
+        Transaction writer = txMgr.runTaskWithRetry(txn -> {
+            put(txn, TEST_TABLE, "row", "column", "first");
+            return txn;
+        });
+
+        assertThat(timestampCache.getCommitTimestampIfPresent(writer.getTimestamp())).isNull();
+
+        txMgr.runTaskWithRetry(txn -> {
+            put(txn, TEST_TABLE, "row", "column", "second");
+            return txn;
+        });
+
+        assertThat(timestampCache.getCommitTimestampIfPresent(writer.getTimestamp()))
+                .isNotNull()
+                .isEqualTo(transactionService.get(writer.getTimestamp()));
+    }
+
+    @Test
+    public void abortedReadTransactionStillCachesWritersCommitTimestamp() {
```

Comment:
nit: maybe `readThenAbortStillCaches...`? A `ReadTransaction` is something else, and could lead to confusion